esphome/components/bl0940/number/__init__.py

```diff
@@ -0,0 +1,93 @@
+import esphome.codegen as cg
+from esphome.components import number
+import esphome.config_validation as cv
+from esphome.const import (
+    CONF_MAX_VALUE,
+    CONF_MIN_VALUE,
+    CONF_MODE,
+    CONF_RESTORE_VALUE,
+    CONF_STEP,
+    ENTITY_CATEGORY_CONFIG,
+    UNIT_PERCENT,
+)
+
+from .. import CONF_BL0940_ID, bl0940_ns
+from ..sensor import BL0940
+
+# Define calibration types
+CONF_CURRENT_CALIBRATION = "current_calibration"
+CONF_VOLTAGE_CALIBRATION = "voltage_calibration"
+CONF_POWER_CALIBRATION = "power_calibration"
+CONF_ENERGY_CALIBRATION = "energy_calibration"
+
+BL0940Number = bl0940_ns.class_("BL0940Number")
+
+CalibrationNumber = bl0940_ns.class_(
+    "CalibrationNumber", number.Number, cg.PollingComponent
+)
+
+
+def validate_min_max(config):
+    if config[CONF_MAX_VALUE] <= config[CONF_MIN_VALUE]:
+        raise cv.Invalid("max_value must be greater than min_value")
+    if config[CONF_MIN_VALUE] < -50 or config[CONF_MAX_VALUE] > 50:
+        raise cv.Invalid("max calibration range is +/-50 percent")
+    return config
+
+
+CALIBRATION_SCHEMA = cv.All(
+    number.number_schema(
+        CalibrationNumber,
+        entity_category=ENTITY_CATEGORY_CONFIG,
+        unit_of_measurement=UNIT_PERCENT,
+    )
+    .extend(
+        {
+            cv.Optional(CONF_MAX_VALUE, default=10): cv.float_,
+            cv.Optional(CONF_MIN_VALUE, default=-10): cv.float_,
+            cv.Optional(CONF_STEP, default=0.1): cv.positive_float,
+            cv.Optional(CONF_RESTORE_VALUE): cv.boolean,
+        }
+    )
+    .extend(cv.polling_component_schema("never")),
+    validate_min_max,
+)
+
+# Configuration schema for BL0940 numbers
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(BL0940Number),
+        cv.Required(CONF_BL0940_ID): cv.use_id(BL0940),
```

Comment:
We can auto link if there is only one
```suggestion
        cv.GenerateID(CONF_BL0940_ID): cv.use_id(BL0940),
```